californium-core/src/main/java/org/eclipse/californium/core/network/EndpointManager.java

```diff
@@ -358,9 +258,12 @@ public void deliverRequest(Exchange exchange) {
 
 		@Override
 		public void deliverResponse(Exchange exchange, Response response) {
-			if (exchange == null) throw new NullPointerException();
-			if (exchange.getRequest() == null) throw new NullPointerException();
-			if (response == null) throw new NullPointerException();
+			if (exchange == null)
+				throw new NullPointerException("no CoAP exchange!");
```

Comment:
In my opinion, the API requires a exchange and a response as precondition. 
Therefore the Exception signalling that something went wrong seems to be the right.
Or what do you have in mind, how should it be used without (accidentley?) ?